pkg/providers/vsphere/session/session_vm_update_test.go

```diff
@@ -1439,25 +1439,79 @@ var _ = Describe("UpdateVirtualMachine", func() {
 					})
 				})
 
-				It("should not add virtual controllers to VM during update when VM is powered on", func() {
+				It("should not edit virtual controllers to VM during update when VM is powered on", func() {
+					By("Set existing nvme controller", func() {
+						vm.Spec.Hardware.NVMEControllers = []vmopv1.NVMEControllerSpec{
+							{
+								BusNumber:   1,
+								SharingMode: vmopv1.VirtualControllerSharingModeNone,
+							},
+						}
+
+						err := sess.UpdateVirtualMachine(vmCtx, vcVM, getUpdateArgs, getResizeArgs)
+						Expect(errors.Is(err, vsphere.ErrReconfigure)).To(BeTrue())
+						// Refresh VM properties
+						Expect(vcVM.Properties(ctx, vcVM.Reference(), vmProps, &vmCtx.MoVM)).To(Succeed())
+					})
+
+					By("Update NVME controller's sharingMode", func() {
+						vm.Spec.Hardware.NVMEControllers = []vmopv1.NVMEControllerSpec{
+							{
+								BusNumber:   1,
+								SharingMode: vmopv1.VirtualControllerSharingModePhysical, // Changed to physical.
+							},
+						}
+					})
+
+					By("Setting VM power state to On", func() {
+						vmCtx.MoVM.Runtime.PowerState = vimtypes.VirtualMachinePowerStatePoweredOn
+					})
+
+					By("Updating the VM", func() {
+						// Update the VM
+						Expect(sess.UpdateVirtualMachine(vmCtx, vcVM, getUpdateArgs, getResizeArgs)).To(Succeed())
+					})
+
+					By("Verifying NVME controller was not updated", func() {
+						Expect(vcVM.Properties(ctx, vcVM.Reference(), vmProps, &vmCtx.MoVM)).To(Succeed())
+						// Collect all the existing controller device keys.
+						nvmeControllerCount := 0
+
+						var nvmeController *vimtypes.VirtualNVMEController
+						fmt.Println(len(vmCtx.MoVM.Config.Hardware.Device))
+						for _, d := range vmCtx.MoVM.Config.Hardware.Device {
```

Comment:
Nit: if you like, this loop can be simplified with `object.VirtualDeviceList(vmCtx.MoVM.Config.Hardware.Device).SelectByType()`